### PR TITLE
Assume all unknown inputs are linker inputs (object files)

### PR DESCRIPTION
**Pull request by sbc100**
This matches the behaviour of clang and gcc.

Fixes: #10495

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -1000,28 +994,23 @@ def add_link_flag(i, f):
                     '-current_version', '-I', '-L', '-include-pch'):
           continue # ignore this gcc-style argument
 
-      if options.expand_symlinks and os.path.islink(arg) and get_file_suffix(os.path.realpath(arg)) in SOURCE_ENDINGS + OBJECT_FILE_ENDINGS + DYNAMICLIB_ENDINGS + ASSEMBLY_ENDINGS + HEADER_ENDINGS:
+      if options.expand_symlinks and os.path.islink(arg):
```
Comment (kripken) [inline emcc.py]: Is there a chance something is a parameter, but also happens to be a link? Like `-O2` happens to be a filename?

File: emcc.py [hunk 2]
```diff
@@ -1000,28 +994,23 @@ def add_link_flag(i, f):
...
+        if file_suffix in SOURCE_ENDINGS + DYNAMICLIB_ENDINGS + HEADER_ENDINGS or shared.Building.is_ar(arg):
+          # we already removed -o <target>, so all these should be inputs
           newargs[i] = ''
           if file_suffix.endswith(SOURCE_ENDINGS):
             input_files.append((i, arg))
             has_source_inputs = True
           elif file_suffix.endswith(HEADER_ENDINGS):
             input_files.append((i, arg))
             has_header_inputs = True
-          elif file_suffix.endswith(ASSEMBLY_ENDINGS) or shared.Building.is_bitcode(arg) or shared.Building.is_ar(arg):
```
Comment (kripken) [inline emcc.py]: the `is_ar` code is no longer present, do we not need it?

**Comment by Cloudef** [comment]
Personal opinion, but if we can make emcc to be more like gcc/clang. That would be very nice in my view. The current `EMMAKEN_JUST_CONFIGURE` and such hacks are bit of confusing.

File: emcc.py (hunk 1 above)
Comment (sbc100) [reply emcc.py]: Probably won't happen but yes, makes sense to move this check below. Done.

File: emcc.py (hunk 2 above)
Comment (sbc100) [reply emcc.py]: Oops.  Thanks.  Restoring.

**Comment by sbc100** [comment]
I'm personally convinced this is the better behaviour.   I can think of no good reason to things any other way.  

The only users who should be effected are those using non-standard source file extensions, but those users will also not be able to build with gcc or clang.

**Comment by sbc100** [comment]
This PR has grown a little .. i'm going to see if I can split it up.

File: emcc.py [hunk 3]
```diff
@@ -1153,8 +1140,8 @@ def check(input_file):
 
     input_files = filter_out_dynamic_libs(input_files)
 
-    if len(input_files) == 0:
-      exit_with_error('no input files\nnote that input files without a known suffix are ignored, make sure your input files end with one of: ' + str(SOURCE_ENDINGS + OBJECT_FILE_ENDINGS + DYNAMICLIB_ENDINGS + STATICLIB_ENDINGS + ASSEMBLY_ENDINGS + HEADER_ENDINGS))
+    if not input_files and not libs:
```
Comment (kripken) [inline emcc.py]: do we have a test for the `not libs` addition here?

**Comment by sbc100** [comment]
OK I'm bringing this change back from the dead :)

**Pull request merged**